<a href="https://colab.research.google.com/github/tuntunwinkic/ML_Specialization/blob/main/content_based.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!pip install pickle5

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 132.1/132.1 kB 1.6 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for pickle5: filename=pickle5-0.0.11-cp310-cp310-linux_x86_64.whl size=255317 sha256=27c21afd21c77a45232710dc2c4ca6bb1176e96eb5c8cf7314fcb6b783f2c8d1
  Stored in directory: /root/.cache/pip/wheels/7d/14/ef/4aab19d27fa8e58772be5c71c16add0426acf9e1f64353235c
Successfully built pickle5


In [ ]:
""" Utilities for RecSysNN assigment """
from collections import defaultdict
import csv
import numpy as np
from numpy import genfromtxt
import pickle5 as pickle
import tabulate


def load_data():
    ''' called to load preprepared data for the lab '''
    item_train = genfromtxt('content_item_train.csv', delimiter=',')
    user_train = genfromtxt('content_user_train.csv', delimiter=',')
    y_train    = genfromtxt('content_y_train.csv', delimiter=',')
    with open('content_item_train_header.txt', newline='') as f:    #csv reader handles quoted strings better
        item_features = list(csv.reader(f))[0]
    with open('content_user_train_header.txt', newline='') as f:
        user_features = list(csv.reader(f))[0]
    item_vecs = genfromtxt('content_item_vecs.csv', delimiter=',')

    movie_dict = defaultdict(dict)
    count = 0
#    with open('./data/movies.csv', newline='') as csvfile:
    with open('content_movie_list.csv', newline='') as csvfile:
        reader = csv.reader(csvfile, delimiter=',', quotechar='"')
        for line in reader:
            if count == 0:
                count += 1  #skip header
                #print(line) print
            else:
                count += 1
                movie_id = int(line[0])
                movie_dict[movie_id]["title"] = line[1]
                movie_dict[movie_id]["genres"] = line[2]

    with open('content_user_to_genre.pickle', 'rb') as f:
        user_to_genre = pickle.load(f)

    return(item_train, user_train, y_train, item_features, user_features, item_vecs, movie_dict, user_to_genre)

def pprint_train(x_train, features, vs, u_s, maxcount=5, user=True):
    """ Prints user_train or item_train nicely """
    if user:
        flist = [".0f", ".0f", ".1f",
                 ".1f", ".1f", ".1f", ".1f", ".1f", ".1f", ".1f", ".1f", ".1f", ".1f", ".1f", ".1f", ".1f", ".1f"]
    else:
        flist = [".0f", ".0f", ".1f",
                 ".0f", ".0f", ".0f", ".0f", ".0f", ".0f", ".0f", ".0f", ".0f", ".0f", ".0f", ".0f", ".0f", ".0f"]

    head = features[:vs]
    if vs < u_s: print("error, vector start {vs} should be greater then user start {u_s}")
    for i in range(u_s):
        head[i] = "[" + head[i] + "]"
    genres = features[vs:]
    hdr = head + genres
    disp = [split_str(hdr, 5)]
    count = 0
    for i in range(0, x_train.shape[0]):
        if count == maxcount: break
        count += 1
        disp.append([x_train[i, 0].astype(int),
                     x_train[i, 1].astype(int),
                     x_train[i, 2].astype(float),
                     *x_train[i, 3:].astype(float)
                    ])
    table = tabulate.tabulate(disp, tablefmt='html', headers="firstrow", floatfmt=flist, numalign='center')
    return table


def split_str(ifeatures, smax):
    ''' split the feature name strings to tables fit '''
    ofeatures = []
    for s in ifeatures:
        if not ' ' in s:  # skip string that already have a space
            if len(s) > smax:
                mid = int(len(s)/2)
                s = s[:mid] + " " + s[mid:]
        ofeatures.append(s)
    return ofeatures


def print_pred_movies(y_p, item, movie_dict, maxcount=10):
    """ print results of prediction of a new user. inputs are expected to be in
        sorted order, unscaled. """
    count = 0
    disp = [["y_p", "movie id", "rating ave", "title", "genres"]]

    for i in range(0, y_p.shape[0]):
        if count == maxcount:
            break
        count += 1
        movie_id = item[i, 0].astype(int)
        disp.append([np.around(y_p[i, 0], 1), item[i, 0].astype(int), np.around(item[i, 2].astype(float), 1),
                     movie_dict[movie_id]['title'], movie_dict[movie_id]['genres']])

    table = tabulate.tabulate(disp, tablefmt='html', headers="firstrow")
    return table

def gen_user_vecs(user_vec, num_items):
    """ given a user vector return:
        user predict maxtrix to match the size of item_vecs """
    user_vecs = np.tile(user_vec, (num_items, 1))
    return user_vecs

# predict on  everything, filter on print/use
def predict_uservec(user_vecs, item_vecs, model, u_s, i_s, scaler):
    """ given a scaled user vector, does the prediction on all movies in scaled print_item_vecs returns
        an array predictions sorted by predicted rating,
        arrays of user and item, sorted by predicted rating sorting index
    """
    y_p = model.predict([user_vecs[:, u_s:], item_vecs[:, i_s:]])
    y_pu = scaler.inverse_transform(y_p)

    if np.any(y_pu < 0):
        print("Error, expected all positive predictions")
    sorted_index = np.argsort(-y_pu, axis=0).reshape(-1).tolist()  #negate to get largest rating first
    sorted_ypu   = y_pu[sorted_index]
    sorted_items = item_vecs[sorted_index]
    sorted_user  = user_vecs[sorted_index]
    return(sorted_index, sorted_ypu, sorted_items, sorted_user)

def get_user_vecs(user_id, user_train, item_vecs, user_to_genre):
    """ given a user_id, return:
        user train/predict matrix to match the size of item_vecs
        y vector with ratings for all rated movies and 0 for others of size item_vecs """

    if not user_id in user_to_genre:
        print("error: unknown user id")
        return None
    else:
        user_vec_found = False
        for i in range(len(user_train)):
            if user_train[i, 0] == user_id:
                user_vec = user_train[i]
                user_vec_found = True
                break
        if not user_vec_found:
            print("error in get_user_vecs, did not find uid in user_train")
        num_items = len(item_vecs)
        user_vecs = np.tile(user_vec, (num_items, 1))

        y = np.zeros(num_items)
        for i in range(num_items):  # walk through movies in item_vecs and get the movies, see if user has rated them
            movie_id = item_vecs[i, 0]
            if movie_id in user_to_genre[user_id]['movies']:
                rating = user_to_genre[user_id]['movies'][movie_id]
            else:
                rating = 0
            y[i] = rating
    return(user_vecs, y)

def get_item_genres(item_gvec, genre_features):
    ''' takes in the item's genre vector and list of genre names
    returns the feature names where gvec was 1 '''
    offsets = np.nonzero(item_gvec)[0]
    genres = [genre_features[i] for i in offsets]
    return genres


def print_existing_user(y_p, y, user, items, ivs, uvs, movie_dict, maxcount=10):
    """ print results of prediction for a user who was in the database.
        Inputs are expected to be in sorted order, unscaled.
    """
    count = 0
    disp = [["y_p", "y", "user", "user genre ave", "movie rating ave", "movie id", "title", "genres"]]
    count = 0
    for i in range(0, y.shape[0]):
        if y[i, 0] != 0:  # zero means not rated
            if count == maxcount:
                break
            count += 1
            movie_id = items[i, 0].astype(int)

            offsets = np.nonzero(items[i, ivs:] == 1)[0]
            genre_ratings = user[i, uvs + offsets]
            disp.append([y_p[i, 0], y[i, 0],
                         user[i, 0].astype(int),      # userid
                         np.array2string(genre_ratings,
                                         formatter={'float_kind':lambda x: "%.1f" % x},
                                         separator=',', suppress_small=True),
                         items[i, 2].astype(float),    # movie average rating
                         movie_id,
                         movie_dict[movie_id]['title'],
                         movie_dict[movie_id]['genres']])

    table = tabulate.tabulate(disp, tablefmt='html', headers="firstrow", floatfmt=[".1f", ".1f", ".0f", ".2f", ".1f"])
    return table


In [ ]:
from tensorflow.keras.activations import relu, linear
from tensorflow.keras.layers import Dense

import numpy as np

def test_tower(target):
    num_outputs = 32
    i = 0
    assert len(target.layers) == 3, f"Wrong number of layers. Expected 3 but got {len(target.layers)}"
    expected = [[Dense, [None, 256], relu],
                [Dense, [None, 128], relu],
                [Dense, [None, num_outputs], linear]]

    for layer in target.layers:
        assert type(layer) == expected[i][0], \
            f"Wrong type in layer {i}. Expected {expected[i][0]} but got {type(layer)}"
        assert layer.output.shape.as_list() == expected[i][1], \
            f"Wrong number of units in layer {i}. Expected {expected[i][1]} but got {layer.output.shape.as_list()}"
        assert layer.activation == expected[i][2], \
            f"Wrong activation in layer {i}. Expected {expected[i][2]} but got {layer.activation}"
        i = i + 1

    print("\033[92mAll tests passed!")


def test_sq_dist(target):
    a1 = np.array([1.0, 2.0, 3.0]); b1 = np.array([1.0, 2.0, 3.0])
    c1 = target(a1, b1)
    a2 = np.array([1.1, 2.1, 3.1]); b2 = np.array([1.0, 2.0, 3.0])
    c2 = target(a2, b2)
    a3 = np.array([0, 1]);          b3 = np.array([1, 0])
    c3 = target(a3, b3)
    a4 = np.array([1, 1, 1, 1, 1]); b4 = np.array([0, 0, 0, 0, 0])
    c4 = target(a4, b4)

    assert np.isclose(c1, 0), f"Wrong value. Expected {0}, got {c1}"
    assert np.isclose(c2, 0.03), f"Wrong value. Expected {0.03}, got {c2}"
    assert np.isclose(c3, 2), f"Wrong value. Expected {2}, got {c3}"
    assert np.isclose(c4, 5), f"Wrong value. Expected {5}, got {c4}"

    print('\033[92mAll tests passed!')


In [ ]:
import numpy as np
import numpy.ma as ma
import pandas as pd
import tensorflow as tf
from tensorflow import keras
from sklearn.preprocessing import StandardScaler, MinMaxScaler
from sklearn.model_selection import train_test_split
import tabulate
pd.set_option("display.precision", 1)

In [ ]:
top10_df = pd.read_csv("content_top10_df.csv")
bygenre_df = pd.read_csv("content_bygenre_df.csv")
top10_df

movie id  num ratings  ave rating  \
0      4993          198         4.1   
1      5952          188         4.0   
2      7153          185         4.1   
3      4306          170         3.9   
4     58559          149         4.2   
5      6539          149         3.8   
6     79132          143         4.1   
7      6377          141         4.0   
8      4886          132         3.9   
9      7361          131         4.2   

                                               title  \
0  Lord of the Rings: The Fellowship of the Ring,...   
1             Lord of the Rings: The Two Towers, The   
2     Lord of the Rings: The Return of the King, The   
3                                              Shrek   
4                                   Dark Knight, The   
5  Pirates of the Caribbean: The Curse of the Bla...   
6                                          Inception   
7                                       Finding Nemo   
8                                     Monsters, Inc.   
9              Eternal Sunshine of the Spotless Mind   

                                              genres  
0                                  Adventure|Fantasy  
1                                  Adventure|Fantasy  
2                     Action|Adventure|Drama|Fantasy  
3  Adventure|Animation|Children|Comedy|Fantasy|Ro...  
4                                 Action|Crime|Drama  
5                    Action|Adventure|Comedy|Fantasy  
6         Action|Crime|Drama|Mystery|Sci-Fi|Thriller  
7                Adventure|Animation|Children|Comedy  
8        Adventure|Animation|Children|Comedy|Fantasy  
9                               Drama|Romance|Sci-Fi

In [ ]:
bygenre_df

genre  num movies  ave rating/genre  ratings per genre
0        Action         321               3.4              10377
1     Adventure         234               3.4               8785
2     Animation          76               3.6               2588
3      Children          69               3.4               2472
4        Comedy         326               3.4               8911
5         Crime         139               3.5               4671
6   Documentary          13               3.8                280
7         Drama         342               3.6              10201
8       Fantasy         124               3.4               4468
9        Horror          56               3.2               1345
10      Mystery          68               3.6               2497
11      Romance         151               3.4               4468
12       Sci-Fi         174               3.4               5894
13     Thriller         245               3.4               7659

In [ ]:
# Load Data, set configuration variables
item_train, user_train, y_train, item_features, user_features, item_vecs, movie_dict, user_to_genre = load_data()

num_user_features = user_train.shape[1] - 3  # remove userid, rating count and ave rating during training
num_item_features = item_train.shape[1] - 1  # remove movie id at train time
uvs = 3  # user genre vector start
ivs = 3  # item genre vector start
u_s = 3  # start of columns to use in training, user
i_s = 1  # start of columns to use in training, items
print(f"Number of training vectors: {len(item_train)}")

Number of training vectors: 50884


In [ ]:
print(item_train.shape)
print(user_train.shape)
print(y_train.shape)
print(len(item_features))
print(len(user_features))

(50884, 17)
(50884, 17)
(50884,)
17
17


In [ ]:
print(item_features)
print(user_features)

['movie id', 'year', 'ave rating', 'Action', 'Adventure', 'Animation', 'Children', 'Comedy', 'Crime', 'Documentary', 'Drama', 'Fantasy', 'Horror', 'Mystery', 'Romance', 'Sci-Fi', 'Thriller']
['user id', 'rating count', 'rating ave', 'Action', 'Adventure', 'Animation', 'Children', 'Comedy', 'Crime', 'Documentary', 'Drama', 'Fantasy', 'Horror', 'Mystery', 'Romance', 'Sci-Fi', 'Thriller']


In [ ]:
pprint_train(user_train, user_features, uvs,  u_s, maxcount=5)

'<table>\n<thead>\n<tr><th style="text-align: center;"> [user id] </th><th style="text-align: center;"> [rating count] </th><th style="text-align: center;"> [rating ave] </th><th style="text-align: center;"> Act ion </th><th style="text-align: center;"> Adve nture </th><th style="text-align: center;"> Anim ation </th><th style="text-align: center;"> Chil dren </th><th style="text-align: center;"> Com edy </th><th style="text-align: center;"> Crime </th><th style="text-align: center;"> Docum entary </th><th style="text-align: center;"> Drama </th><th style="text-align: center;"> Fan tasy </th><th style="text-align: center;"> Hor ror </th><th style="text-align: center;"> Mys tery </th><th style="text-align: center;"> Rom ance </th><th style="text-align: center;"> Sci -Fi </th><th style="text-align: center;"> Thri ller </th></tr>\n</thead>\n<tbody>\n<tr><td style="text-align: center;">     2     </td><td style="text-align: center;">       22       </td><td style="text-align: center;">     4.0      </td><td style="text-align: center;">   4.0   </td><td style="text-align: center;">    4.2     </td><td style="text-align: center;">    0.0     </td><td style="text-align: center;">    0.0    </td><td style="text-align: center;">   4.0   </td><td style="text-align: center;">  4.1  </td><td style="text-align: center;">     4.0      </td><td style="text-align: center;">  4.0  </td><td style="text-align: center;">   0.0    </td><td style="text-align: center;">   3.0   </td><td style="text-align: center;">   4.0    </td><td style="text-align: center;">   0.0    </td><td style="text-align: center;">   3.9   </td><td style="text-align: center;">    3.9    </td></tr>\n<tr><td style="text-align: center;">     2     </td><td style="text-align: center;">       22       </td><td style="text-align: center;">     4.0      </td><td style="text-align: center;">   4.0   </td><td style="text-align: center;">    4.2     </td><td style="text-align: center;">    0.0     </td><td style="text-align: center;">    0.0    </td><td style="text-align: center;">   4.0   </td><td style="text-align: center;">  4.1  </td><td style="text-align: center;">     4.0      </td><td style="text-align: center;">  4.0  </td><td style="text-align: center;">   0.0    </td><td style="text-align: center;">   3.0   </td><td style="text-align: center;">   4.0    </td><td style="text-align: center;">   0.0    </td><td style="text-align: center;">   3.9   </td><td style="text-align: center;">    3.9    </td></tr>\n<tr><td style="text-align: center;">     2     </td><td style="text-align: center;">       22       </td><td style="text-align: center;">     4.0      </td><td style="text-align: center;">   4.0   </td><td style="text-align: center;">    4.2     </td><td style="text-align: center;">    0.0     </td><td style="text-align: center;">    0.0    </td><td style="text-align: center;">   4.0   </td><td style="text-align: center;">  4.1  </td><td style="text-align: center;">     4.0      </td><td style="text-align: center;">  4.0  </td><td style="text-align: center;">   0.0    </td><td style="text-align: center;">   3.0   </td><td style="text-align: center;">   4.0    </td><td style="text-align: center;">   0.0    </td><td style="text-align: center;">   3.9   </td><td style="text-align: center;">    3.9    </td></tr>\n<tr><td style="text-align: center;">     2     </td><td style="text-align: center;">       22       </td><td style="text-align: center;">     4.0      </td><td style="text-align: center;">   4.0   </td><td style="text-align: center;">    4.2     </td><td style="text-align: center;">    0.0     </td><td style="text-align: center;">    0.0    </td><td style="text-align: center;">   4.0   </td><td style="text-align: center;">  4.1  </td><td style="text-align: center;">     4.0      </td><td style="text-align: center;">  4.0  </td><td style="text-align: center;">   0.0    </td><td style="text-align: center;">   3.0   </td><td style="text-align: center;">   4.0    </td><td s

<table>
<thead>
<tr><th style="text-align: center;"> [user id] </th><th style="text-align: center;"> [rating count] </th><th style="text-align: center;"> [rating ave] </th><th style="text-align: center;"> Act ion </th><th style="text-align: center;"> Adve nture </th><th style="text-align: center;"> Anim ation </th><th style="text-align: center;"> Chil dren </th><th style="text-align: center;"> Com edy </th><th style="text-align: center;"> Crime </th><th style="text-align: center;"> Docum entary </th><th style="text-align: center;"> Drama </th><th style="text-align: center;"> Fan tasy </th><th style="text-align: center;"> Hor ror </th><th style="text-align: center;"> Mys tery </th><th style="text-align: center;"> Rom ance </th><th style="text-align: center;"> Sci -Fi </th><th style="text-align: center;"> Thri ller </th></tr>
</thead>
<tbody>
<tr><td style="text-align: center;">     2     </td><td style="text-align: center;">       22       </td><td style="text-align: center;">     4.0      </td><td style="text-align: center;">   4.0   </td><td style="text-align: center;">    4.2     </td><td style="text-align: center;">    0.0     </td><td style="text-align: center;">    0.0    </td><td style="text-align: center;">   4.0   </td><td style="text-align: center;">  4.1  </td><td style="text-align: center;">     4.0      </td><td style="text-align: center;">  4.0  </td><td style="text-align: center;">   0.0    </td><td style="text-align: center;">   3.0   </td><td style="text-align: center;">   4.0    </td><td style="text-align: center;">   0.0    </td><td style="text-align: center;">   3.9   </td><td style="text-align: center;">    3.9    </td></tr>
<tr><td style="text-align: center;">     2     </td><td style="text-align: center;">       22       </td><td style="text-align: center;">     4.0      </td><td style="text-align: center;">   4.0   </td><td style="text-align: center;">    4.2     </td><td style="text-align: center;">    0.0     </td><td style="text-align: center;">    0.0    </td><td style="text-align: center;">   4.0   </td><td style="text-align: center;">  4.1  </td><td style="text-align: center;">     4.0      </td><td style="text-align: center;">  4.0  </td><td style="text-align: center;">   0.0    </td><td style="text-align: center;">   3.0   </td><td style="text-align: center;">   4.0    </td><td style="text-align: center;">   0.0    </td><td style="text-align: center;">   3.9   </td><td style="text-align: center;">    3.9    </td></tr>
<tr><td style="text-align: center;">     2     </td><td style="text-align: center;">       22       </td><td style="text-align: center;">     4.0      </td><td style="text-align: center;">   4.0   </td><td style="text-align: center;">    4.2     </td><td style="text-align: center;">    0.0     </td><td style="text-align: center;">    0.0    </td><td style="text-align: center;">   4.0   </td><td style="text-align: center;">  4.1  </td><td style="text-align: center;">     4.0      </td><td style="text-align: center;">  4.0  </td><td style="text-align: center;">   0.0    </td><td style="text-align: center;">   3.0   </td><td style="text-align: center;">   4.0    </td><td style="text-align: center;">   0.0    </td><td style="text-align: center;">   3.9   </td><td style="text-align: center;">    3.9    </td></tr>
<tr><td style="text-align: center;">     2     </td><td style="text-align: center;">       22       </td><td style="text-align: center;">     4.0      </td><td style="text-align: center;">   4.0   </td><td style="text-align: center;">    4.2     </td><td style="text-align: center;">    0.0     </td><td style="text-align: center;">    0.0    </td><td style="text-align: center;">   4.0   </td><td style="text-align: center;">  4.1  </td><td style="text-align: center;">     4.0      </td><td style="text-align: center;">  4.0  </td><td style="text-align: center;">   0.0    </td><td style="text-align: center;">   3.0   </td><td style="text-align: center;">   4.0    </td><td style="text-align: center;">   0.0    </td><td style="text-align: center;">   3.9   </td><td style="text-align: center;">    3.9    </td></tr>
<tr><td style="text-align: center;">     2     </td><td style="text-align: center;">       22       </td><td style="text-align: center;">     4.0      </td><td style="text-align: center;">   4.0   </td><td style="text-align: center;">    4.2     </td><td style="text-align: center;">    0.0     </td><td style="text-align: center;">    0.0    </td><td style="text-align: center;">   4.0   </td><td style="text-align: center;">  4.1  </td><td style="text-align: center;">     4.0      </td><td style="text-align: center;">  4.0  </td><td style="text-align: center;">   0.0    </td><td style="text-align: center;">   3.0   </td><td style="text-align: center;">   4.0    </td><td style="text-align: center;">   0.0    </td><td style="text-align: center;">   3.9   </td><td style="text-align: center;">    3.9    </td></tr>
</tbody>
</table>

In [ ]:
pprint_train(item_train, item_features, ivs, i_s, maxcount=5, user=False)

'<table>\n<thead>\n<tr><th style="text-align: center;"> [movie id] </th><th style="text-align: center;"> year </th><th style="text-align: center;"> ave rating </th><th style="text-align: center;"> Act ion </th><th style="text-align: center;"> Adve nture </th><th style="text-align: center;"> Anim ation </th><th style="text-align: center;"> Chil dren </th><th style="text-align: center;"> Com edy </th><th style="text-align: center;"> Crime </th><th style="text-align: center;"> Docum entary </th><th style="text-align: center;"> Drama </th><th style="text-align: center;"> Fan tasy </th><th style="text-align: center;"> Hor ror </th><th style="text-align: center;"> Mys tery </th><th style="text-align: center;"> Rom ance </th><th style="text-align: center;"> Sci -Fi </th><th style="text-align: center;"> Thri ller </th></tr>\n</thead>\n<tbody>\n<tr><td style="text-align: center;">    6874    </td><td style="text-align: center;"> 2003 </td><td style="text-align: center;">    4.0     </td><td style="text-align: center;">    1    </td><td style="text-align: center;">     0      </td><td style="text-align: center;">     0      </td><td style="text-align: center;">     0     </td><td style="text-align: center;">    0    </td><td style="text-align: center;">   1   </td><td style="text-align: center;">      0       </td><td style="text-align: center;">   0   </td><td style="text-align: center;">    0     </td><td style="text-align: center;">    0    </td><td style="text-align: center;">    0     </td><td style="text-align: center;">    0     </td><td style="text-align: center;">    0    </td><td style="text-align: center;">     1     </td></tr>\n<tr><td style="text-align: center;">    8798    </td><td style="text-align: center;"> 2004 </td><td style="text-align: center;">    3.8     </td><td style="text-align: center;">    1    </td><td style="text-align: center;">     0      </td><td style="text-align: center;">     0      </td><td style="text-align: center;">     0     </td><td style="text-align: center;">    0    </td><td style="text-align: center;">   1   </td><td style="text-align: center;">      0       </td><td style="text-align: center;">   1   </td><td style="text-align: center;">    0     </td><td style="text-align: center;">    0    </td><td style="text-align: center;">    0     </td><td style="text-align: center;">    0     </td><td style="text-align: center;">    0    </td><td style="text-align: center;">     1     </td></tr>\n<tr><td style="text-align: center;">   46970    </td><td style="text-align: center;"> 2006 </td><td style="text-align: center;">    3.2     </td><td style="text-align: center;">    1    </td><td style="text-align: center;">     0      </td><td style="text-align: center;">     0      </td><td style="text-align: center;">     0     </td><td style="text-align: center;">    1    </td><td style="text-align: center;">   0   </td><td style="text-align: center;">      0       </td><td style="text-align: center;">   0   </td><td style="text-align: center;">    0     </td><td style="text-align: center;">    0    </td><td style="text-align: center;">    0     </td><td style="text-align: center;">    0     </td><td style="text-align: center;">    0    </td><td style="text-align: center;">     0     </td></tr>\n<tr><td style="text-align: center;">   48516    </td><td style="text-align: center;"> 2006 </td><td style="text-align: center;">    4.3     </td><td style="text-align: center;">    0    </td><td style="text-align: center;">     0      </td><td style="text-align: center;">     0      </td><td style="text-align: center;">     0     </td><td style="text-align: center;">    0    </td><td style="text-align: center;">   1   </td><td style="text-align: center;">      0       </td><td style="text-align: center;">   1   </td><td style="text-align: center;">    0     </td><td style="text-align: center;">    0    </td><td style="text-align: center;">    0     </td><td style="text-align: center;">    0     </td><td style="te

<table>
<thead>
<tr><th style="text-align: center;"> [movie id] </th><th style="text-align: center;"> year </th><th style="text-align: center;"> ave rating </th><th style="text-align: center;"> Act ion </th><th style="text-align: center;"> Adve nture </th><th style="text-align: center;"> Anim ation </th><th style="text-align: center;"> Chil dren </th><th style="text-align: center;"> Com edy </th><th style="text-align: center;"> Crime </th><th style="text-align: center;"> Docum entary </th><th style="text-align: center;"> Drama </th><th style="text-align: center;"> Fan tasy </th><th style="text-align: center;"> Hor ror </th><th style="text-align: center;"> Mys tery </th><th style="text-align: center;"> Rom ance </th><th style="text-align: center;"> Sci -Fi </th><th style="text-align: center;"> Thri ller </th></tr>
</thead>
<tbody>
<tr><td style="text-align: center;">    6874    </td><td style="text-align: center;"> 2003 </td><td style="text-align: center;">    4.0     </td><td style="text-align: center;">    1    </td><td style="text-align: center;">     0      </td><td style="text-align: center;">     0      </td><td style="text-align: center;">     0     </td><td style="text-align: center;">    0    </td><td style="text-align: center;">   1   </td><td style="text-align: center;">      0       </td><td style="text-align: center;">   0   </td><td style="text-align: center;">    0     </td><td style="text-align: center;">    0    </td><td style="text-align: center;">    0     </td><td style="text-align: center;">    0     </td><td style="text-align: center;">    0    </td><td style="text-align: center;">     1     </td></tr>
<tr><td style="text-align: center;">    8798    </td><td style="text-align: center;"> 2004 </td><td style="text-align: center;">    3.8     </td><td style="text-align: center;">    1    </td><td style="text-align: center;">     0      </td><td style="text-align: center;">     0      </td><td style="text-align: center;">     0     </td><td style="text-align: center;">    0    </td><td style="text-align: center;">   1   </td><td style="text-align: center;">      0       </td><td style="text-align: center;">   1   </td><td style="text-align: center;">    0     </td><td style="text-align: center;">    0    </td><td style="text-align: center;">    0     </td><td style="text-align: center;">    0     </td><td style="text-align: center;">    0    </td><td style="text-align: center;">     1     </td></tr>
<tr><td style="text-align: center;">   46970    </td><td style="text-align: center;"> 2006 </td><td style="text-align: center;">    3.2     </td><td style="text-align: center;">    1    </td><td style="text-align: center;">     0      </td><td style="text-align: center;">     0      </td><td style="text-align: center;">     0     </td><td style="text-align: center;">    1    </td><td style="text-align: center;">   0   </td><td style="text-align: center;">      0       </td><td style="text-align: center;">   0   </td><td style="text-align: center;">    0     </td><td style="text-align: center;">    0    </td><td style="text-align: center;">    0     </td><td style="text-align: center;">    0     </td><td style="text-align: center;">    0    </td><td style="text-align: center;">     0     </td></tr>
<tr><td style="text-align: center;">   48516    </td><td style="text-align: center;"> 2006 </td><td style="text-align: center;">    4.3     </td><td style="text-align: center;">    0    </td><td style="text-align: center;">     0      </td><td style="text-align: center;">     0      </td><td style="text-align: center;">     0     </td><td style="text-align: center;">    0    </td><td style="text-align: center;">   1   </td><td style="text-align: center;">      0       </td><td style="text-align: center;">   1   </td><td style="text-align: center;">    0     </td><td style="text-align: center;">    0    </td><td style="text-align: center;">    0     </td><td style="text-align: center;">    0     </td><td style="text-align: center;">    0    </td><td style="text-align: center;">     1     </td></tr>
<tr><td style="text-align: center;">   58559    </td><td style="text-align: center;"> 2008 </td><td style="text-align: center;">    4.2     </td><td style="text-align: center;">    1    </td><td style="text-align: center;">     0      </td><td style="text-align: center;">     0      </td><td style="text-align: center;">     0     </td><td style="text-align: center;">    0    </td><td style="text-align: center;">   1   </td><td style="text-align: center;">      0       </td><td style="text-align: center;">   1   </td><td style="text-align: center;">    0     </td><td style="text-align: center;">    0    </td><td style="text-align: center;">    0     </td><td style="text-align: center;">    0     </td><td style="text-align: center;">    0    </td><td style="text-align: center;">     0     </td></tr>
</tbody>
</table>

In [ ]:
print(f"y_train[:5]: {y_train[:5]}")

y_train[:5]: [4.  3.5 4.  4.  4.5]


In [ ]:
# scale training data
item_train_unscaled = item_train
user_train_unscaled = user_train
y_train_unscaled    = y_train

scalerItem = StandardScaler()
scalerItem.fit(item_train)
item_train = scalerItem.transform(item_train)

scalerUser = StandardScaler()
scalerUser.fit(user_train)
user_train = scalerUser.transform(user_train)

scalerTarget = MinMaxScaler((-1, 1))
scalerTarget.fit(y_train.reshape(-1, 1))
y_train = scalerTarget.transform(y_train.reshape(-1, 1))
#ynorm_test = scalerTarget.transform(y_test.reshape(-1, 1))

print(np.allclose(item_train_unscaled, scalerItem.inverse_transform(item_train)))
print(np.allclose(user_train_unscaled, scalerUser.inverse_transform(user_train)))

True
True


In [ ]:
item_train, item_test = train_test_split(item_train, train_size=0.80, shuffle=True, random_state=1)
user_train, user_test = train_test_split(user_train, train_size=0.80, shuffle=True, random_state=1)
y_train, y_test       = train_test_split(y_train,    train_size=0.80, shuffle=True, random_state=1)
print(f"movie/item training data shape: {item_train.shape}")
print(f"movie/item test data shape: {item_test.shape}")

movie/item training data shape: (40707, 17)
movie/item test data shape: (10177, 17)


In [ ]:
pprint_train(user_train, user_features, uvs, u_s, maxcount=5)

'<table>\n<thead>\n<tr><th style="text-align: center;"> [user id] </th><th style="text-align: center;"> [rating count] </th><th style="text-align: center;"> [rating ave] </th><th style="text-align: center;"> Act ion </th><th style="text-align: center;"> Adve nture </th><th style="text-align: center;"> Anim ation </th><th style="text-align: center;"> Chil dren </th><th style="text-align: center;"> Com edy </th><th style="text-align: center;"> Crime </th><th style="text-align: center;"> Docum entary </th><th style="text-align: center;"> Drama </th><th style="text-align: center;"> Fan tasy </th><th style="text-align: center;"> Hor ror </th><th style="text-align: center;"> Mys tery </th><th style="text-align: center;"> Rom ance </th><th style="text-align: center;"> Sci -Fi </th><th style="text-align: center;"> Thri ller </th></tr>\n</thead>\n<tbody>\n<tr><td style="text-align: center;">     1     </td><td style="text-align: center;">       0        </td><td style="text-align: center;">     -1.0     </td><td style="text-align: center;">  -0.8   </td><td style="text-align: center;">    -0.7    </td><td style="text-align: center;">    0.1     </td><td style="text-align: center;">   -0.0    </td><td style="text-align: center;">  -1.2   </td><td style="text-align: center;"> -0.4  </td><td style="text-align: center;">     0.6      </td><td style="text-align: center;"> -0.5  </td><td style="text-align: center;">   -0.5   </td><td style="text-align: center;">  -0.1   </td><td style="text-align: center;">   -0.6   </td><td style="text-align: center;">   -0.6   </td><td style="text-align: center;">  -0.7   </td><td style="text-align: center;">   -0.7    </td></tr>\n<tr><td style="text-align: center;">     0     </td><td style="text-align: center;">       1        </td><td style="text-align: center;">     -0.7     </td><td style="text-align: center;">  -0.5   </td><td style="text-align: center;">    -0.7    </td><td style="text-align: center;">    -0.1    </td><td style="text-align: center;">   -0.2    </td><td style="text-align: center;">  -0.6   </td><td style="text-align: center;"> -0.2  </td><td style="text-align: center;">     0.7      </td><td style="text-align: center;"> -0.5  </td><td style="text-align: center;">   -0.8   </td><td style="text-align: center;">   0.1   </td><td style="text-align: center;">   -0.0   </td><td style="text-align: center;">   -0.6   </td><td style="text-align: center;">  -0.5   </td><td style="text-align: center;">   -0.4    </td></tr>\n<tr><td style="text-align: center;">    -1     </td><td style="text-align: center;">       -1       </td><td style="text-align: center;">     -0.2     </td><td style="text-align: center;">   0.3   </td><td style="text-align: center;">    -0.4    </td><td style="text-align: center;">    0.4     </td><td style="text-align: center;">    0.5    </td><td style="text-align: center;">   1.0   </td><td style="text-align: center;">  0.6  </td><td style="text-align: center;">     -1.2     </td><td style="text-align: center;"> -0.3  </td><td style="text-align: center;">   -0.6   </td><td style="text-align: center;">  -2.3   </td><td style="text-align: center;">   -0.1   </td><td style="text-align: center;">   0.0    </td><td style="text-align: center;">   0.4   </td><td style="text-align: center;">   -0.0    </td></tr>\n<tr><td style="text-align: center;">     0     </td><td style="text-align: center;">       -1       </td><td style="text-align: center;">     0.6      </td><td style="text-align: center;">   0.5   </td><td style="text-align: center;">    0.5     </td><td style="text-align: center;">    0.2     </td><td style="text-align: center;">    0.6    </td><td style="text-align: center;">  -0.1   </td><td style="text-align: center;">  0.5  </td><td style="text-align: center;">     -1.2     </td><td style="text-align: center;">  0.9  </td><td style="text-align: center;">   1.2    </td><td style="text-align: center;">  -2.3   </td><td style="text-align: center;">   -0.1   </td><td s

<table>
<thead>
<tr><th style="text-align: center;"> [user id] </th><th style="text-align: center;"> [rating count] </th><th style="text-align: center;"> [rating ave] </th><th style="text-align: center;"> Act ion </th><th style="text-align: center;"> Adve nture </th><th style="text-align: center;"> Anim ation </th><th style="text-align: center;"> Chil dren </th><th style="text-align: center;"> Com edy </th><th style="text-align: center;"> Crime </th><th style="text-align: center;"> Docum entary </th><th style="text-align: center;"> Drama </th><th style="text-align: center;"> Fan tasy </th><th style="text-align: center;"> Hor ror </th><th style="text-align: center;"> Mys tery </th><th style="text-align: center;"> Rom ance </th><th style="text-align: center;"> Sci -Fi </th><th style="text-align: center;"> Thri ller </th></tr>
</thead>
<tbody>
<tr><td style="text-align: center;">     1     </td><td style="text-align: center;">       0        </td><td style="text-align: center;">     -1.0     </td><td style="text-align: center;">  -0.8   </td><td style="text-align: center;">    -0.7    </td><td style="text-align: center;">    0.1     </td><td style="text-align: center;">   -0.0    </td><td style="text-align: center;">  -1.2   </td><td style="text-align: center;"> -0.4  </td><td style="text-align: center;">     0.6      </td><td style="text-align: center;"> -0.5  </td><td style="text-align: center;">   -0.5   </td><td style="text-align: center;">  -0.1   </td><td style="text-align: center;">   -0.6   </td><td style="text-align: center;">   -0.6   </td><td style="text-align: center;">  -0.7   </td><td style="text-align: center;">   -0.7    </td></tr>
<tr><td style="text-align: center;">     0     </td><td style="text-align: center;">       1        </td><td style="text-align: center;">     -0.7     </td><td style="text-align: center;">  -0.5   </td><td style="text-align: center;">    -0.7    </td><td style="text-align: center;">    -0.1    </td><td style="text-align: center;">   -0.2    </td><td style="text-align: center;">  -0.6   </td><td style="text-align: center;"> -0.2  </td><td style="text-align: center;">     0.7      </td><td style="text-align: center;"> -0.5  </td><td style="text-align: center;">   -0.8   </td><td style="text-align: center;">   0.1   </td><td style="text-align: center;">   -0.0   </td><td style="text-align: center;">   -0.6   </td><td style="text-align: center;">  -0.5   </td><td style="text-align: center;">   -0.4    </td></tr>
<tr><td style="text-align: center;">    -1     </td><td style="text-align: center;">       -1       </td><td style="text-align: center;">     -0.2     </td><td style="text-align: center;">   0.3   </td><td style="text-align: center;">    -0.4    </td><td style="text-align: center;">    0.4     </td><td style="text-align: center;">    0.5    </td><td style="text-align: center;">   1.0   </td><td style="text-align: center;">  0.6  </td><td style="text-align: center;">     -1.2     </td><td style="text-align: center;"> -0.3  </td><td style="text-align: center;">   -0.6   </td><td style="text-align: center;">  -2.3   </td><td style="text-align: center;">   -0.1   </td><td style="text-align: center;">   0.0    </td><td style="text-align: center;">   0.4   </td><td style="text-align: center;">   -0.0    </td></tr>
<tr><td style="text-align: center;">     0     </td><td style="text-align: center;">       -1       </td><td style="text-align: center;">     0.6      </td><td style="text-align: center;">   0.5   </td><td style="text-align: center;">    0.5     </td><td style="text-align: center;">    0.2     </td><td style="text-align: center;">    0.6    </td><td style="text-align: center;">  -0.1   </td><td style="text-align: center;">  0.5  </td><td style="text-align: center;">     -1.2     </td><td style="text-align: center;">  0.9  </td><td style="text-align: center;">   1.2    </td><td style="text-align: center;">  -2.3   </td><td style="text-align: center;">   -0.1   </td><td style="text-align: center;">   0.0    </td><td style="text-align: center;">   0.2   </td><td style="text-align: center;">    0.3    </td></tr>
<tr><td style="text-align: center;">    -1     </td><td style="text-align: center;">       0        </td><td style="text-align: center;">     0.7      </td><td style="text-align: center;">   0.6   </td><td style="text-align: center;">    0.5     </td><td style="text-align: center;">    0.3     </td><td style="text-align: center;">    0.5    </td><td style="text-align: center;">   0.4   </td><td style="text-align: center;">  0.6  </td><td style="text-align: center;">     1.0      </td><td style="text-align: center;">  0.6  </td><td style="text-align: center;">   0.3    </td><td style="text-align: center;">   0.8   </td><td style="text-align: center;">   0.8    </td><td style="text-align: center;">   0.4    </td><td style="text-align: center;">   0.7   </td><td style="text-align: center;">    0.7    </td></tr>
</tbody>
</table>

In [ ]:
# GRADED_CELL
# UNQ_C1

num_outputs = 32
tf.random.set_seed(1)
user_NN = tf.keras.models.Sequential([
    ### START CODE HERE ###
    tf.keras.layers.Dense(256, activation='relu'),
    tf.keras.layers.Dense(128, activation='relu'),
    tf.keras.layers.Dense(num_outputs),



    ### END CODE HERE ###
])

item_NN = tf.keras.models.Sequential([
    ### START CODE HERE ###
    tf.keras.layers.Dense(256, activation='relu'),
    tf.keras.layers.Dense(128, activation='relu'),
    tf.keras.layers.Dense(num_outputs),


    ### END CODE HERE ###
])

# create the user input and point to the base network
input_user = tf.keras.layers.Input(shape=(num_user_features))
vu = user_NN(input_user)
vu = tf.linalg.l2_normalize(vu, axis=1)

# create the item input and point to the base network
input_item = tf.keras.layers.Input(shape=(num_item_features))
vm = item_NN(input_item)
vm = tf.linalg.l2_normalize(vm, axis=1)

# compute the dot product of the two vectors vu and vm
output = tf.keras.layers.Dot(axes=1)([vu, vm])

# specify the inputs and output of the model
model = tf.keras.Model([input_user, input_item], output)

model.summary()

Model: "model"
__________________________________________________________________________________________________
 Layer (type)                   Output Shape         Param #     Connected to                     
 input_1 (InputLayer)           [(None, 14)]         0           []                               
                                                                                                  
 input_2 (InputLayer)           [(None, 16)]         0           []                               
                                                                                                  
 sequential (Sequential)        (None, 32)           40864       ['input_1[0][0]']                
                                                                                                  
 sequential_1 (Sequential)      (None, 32)           41376       ['input_2[0][0]']                
                                                                                              

In [ ]:
# Public tests
test_tower(user_NN)
test_tower(item_NN)

All tests passed!
All tests passed!


In [ ]:
tf.random.set_seed(1)
cost_fn = tf.keras.losses.MeanSquaredError()
opt = keras.optimizers.Adam(learning_rate=0.01)
model.compile(optimizer=opt,
              loss=cost_fn)

In [ ]:
tf.random.set_seed(1)
model.fit([user_train[:, u_s:], item_train[:, i_s:]], y_train, epochs=30)

Epoch 1/30
1273/1273 [==============================] - 6s 4ms/step - loss: 0.1241
Epoch 2/30
1273/1273 [==============================] - 4s 3ms/step - loss: 0.1148
Epoch 3/30
1273/1273 [==============================] - 4s 3ms/step - loss: 0.1096
Epoch 4/30
1273/1273 [==============================] - 5s 4ms/step - loss: 0.1054
Epoch 5/30
1273/1273 [==============================] - 4s 3ms/step - loss: 0.1029
Epoch 6/30
1273/1273 [==============================] - 4s 3ms/step - loss: 0.1001
Epoch 7/30
1273/1273 [==============================] - 4s 3ms/step - loss: 0.0979
Epoch 8/30
1273/1273 [==============================] - 4s 3ms/step - loss: 0.0959
Epoch 9/30
1273/1273 [==============================] - 4s 3ms/step - loss: 0.0938
Epoch 10/30
1273/1273 [==============================] - 5s 4ms/step - loss: 0.0920
Epoch 11/30
1273/1273 [==============================] - 4s 3ms/step - loss: 0.0900
Epoch 12/30
1273/1273 [==============================] - 4s 3ms/step - loss: 0.0881
E

In [ ]:
model.evaluate([user_test[:, u_s:], item_test[:, i_s:]], y_test)

319/319 [==============================] - 1s 1ms/step - loss: 0.0811


0.0810687392950058

In [ ]:
new_user_id = 5000
new_rating_ave = 0.0
new_action = 0.0
new_adventure = 5.0
new_animation = 0.0
new_childrens = 0.0
new_comedy = 0.0
new_crime = 0.0
new_documentary = 0.0
new_drama = 0.0
new_fantasy = 5.0
new_horror = 0.0
new_mystery = 0.0
new_romance = 0.0
new_scifi = 0.0
new_thriller = 0.0
new_rating_count = 3

user_vec = np.array([[new_user_id, new_rating_count, new_rating_ave,
                      new_action, new_adventure, new_animation, new_childrens,
                      new_comedy, new_crime, new_documentary,
                      new_drama, new_fantasy, new_horror, new_mystery,
                      new_romance, new_scifi, new_thriller]])

In [ ]:
# generate and replicate the user vector to match the number movies in the data set.
user_vecs = gen_user_vecs(user_vec,len(item_vecs))

# scale our user and item vectors
suser_vecs = scalerUser.transform(user_vecs)
sitem_vecs = scalerItem.transform(item_vecs)

# make a prediction
y_p = model.predict([suser_vecs[:, u_s:], sitem_vecs[:, i_s:]])

# unscale y prediction
y_pu = scalerTarget.inverse_transform(y_p)

# sort the results, highest prediction first
sorted_index = np.argsort(-y_pu,axis=0).reshape(-1).tolist()  #negate to get largest rating first
sorted_ypu   = y_pu[sorted_index]
sorted_items = item_vecs[sorted_index]  #using unscaled vectors for display

print_pred_movies(sorted_ypu, sorted_items, movie_dict, maxcount = 10)

27/27 [==============================] - 0s 1ms/step


'<table>\n<thead>\n<tr><th style="text-align: right;">  y_p</th><th style="text-align: right;">  movie id</th><th style="text-align: right;">  rating ave</th><th>title                                              </th><th>genres                           </th></tr>\n</thead>\n<tbody>\n<tr><td style="text-align: right;">  4.3</td><td style="text-align: right;">     54001</td><td style="text-align: right;">         3.9</td><td>Harry Potter and the Order of the Phoenix (2007)   </td><td>Adventure|Drama|Fantasy          </td></tr>\n<tr><td style="text-align: right;">  4.3</td><td style="text-align: right;">     81834</td><td style="text-align: right;">         4  </td><td>Harry Potter and the Deathly Hallows: Part 1 (2010)</td><td>Action|Adventure|Fantasy         </td></tr>\n<tr><td style="text-align: right;">  4.2</td><td style="text-align: right;">    137857</td><td style="text-align: right;">         3.6</td><td>The Jungle Book (2016)                             </td><td>Adventure|Drama|Fantasy          </td></tr>\n<tr><td style="text-align: right;">  4.2</td><td style="text-align: right;">     59387</td><td style="text-align: right;">         4  </td><td>Fall, The (2006)                                   </td><td>Adventure|Drama|Fantasy          </td></tr>\n<tr><td style="text-align: right;">  4.2</td><td style="text-align: right;">      8368</td><td style="text-align: right;">         3.9</td><td>Harry Potter and the Prisoner of Azkaban (2004)    </td><td>Adventure|Fantasy                </td></tr>\n<tr><td style="text-align: right;">  4.2</td><td style="text-align: right;">     57640</td><td style="text-align: right;">         3.8</td><td>Hellboy II: The Golden Army (2008)                 </td><td>Action|Adventure|Fantasy|Sci-Fi  </td></tr>\n<tr><td style="text-align: right;">  4.2</td><td style="text-align: right;">     98809</td><td style="text-align: right;">         3.8</td><td>Hobbit: An Unexpected Journey, The (2012)          </td><td>Adventure|Fantasy                </td></tr>\n<tr><td style="text-align: right;">  4.1</td><td style="text-align: right;">     69844</td><td style="text-align: right;">         3.9</td><td>Harry Potter and the Half-Blood Prince (2009)      </td><td>Adventure|Fantasy|Mystery|Romance</td></tr>\n<tr><td style="text-align: right;">  4.1</td><td style="text-align: right;">      5952</td><td style="text-align: right;">         4  </td><td>Lord of the Rings: The Two Towers, The (2002)      </td><td>Adventure|Fantasy                </td></tr>\n<tr><td style="text-align: right;">  4.1</td><td style="text-align: right;">     54881</td><td style="text-align: right;">         3.9</td><td>King of Kong, The (2007)                           </td><td>Documentary                      </td></tr>\n</tbody>\n</table>'

In [ ]:
uid = 2
# form a set of user vectors. This is the same vector, transformed and repeated.
user_vecs, y_vecs = get_user_vecs(uid, user_train_unscaled, item_vecs, user_to_genre)

# scale our user and item vectors
suser_vecs = scalerUser.transform(user_vecs)
sitem_vecs = scalerItem.transform(item_vecs)

# make a prediction
y_p = model.predict([suser_vecs[:, u_s:], sitem_vecs[:, i_s:]])

# unscale y prediction
y_pu = scalerTarget.inverse_transform(y_p)

# sort the results, highest prediction first
sorted_index = np.argsort(-y_pu,axis=0).reshape(-1).tolist()  #negate to get largest rating first
sorted_ypu   = y_pu[sorted_index]
sorted_items = item_vecs[sorted_index]  #using unscaled vectors for display
sorted_user  = user_vecs[sorted_index]
sorted_y     = y_vecs[sorted_index]

#print sorted predictions for movies rated by the user
print_existing_user(sorted_ypu, sorted_y.reshape(-1,1), sorted_user, sorted_items, ivs, uvs, movie_dict, maxcount = 50)

27/27 [==============================] - 0s 2ms/step


'<table>\n<thead>\n<tr><th style="text-align: right;">  y_p</th><th style="text-align: right;">  y</th><th style="text-align: right;">  user</th><th>user genre ave           </th><th style="text-align: right;">  movie rating ave</th><th style="text-align: right;">  movie id</th><th>title                                             </th><th>genres                                    </th></tr>\n</thead>\n<tbody>\n<tr><td style="text-align: right;">  4.4</td><td style="text-align: right;">5.0</td><td style="text-align: right;">     2</td><td>[4.0]                    </td><td style="text-align: right;">               4.3</td><td style="text-align: right;">     80906</td><td>Inside Job (2010)                                 </td><td>Documentary                               </td></tr>\n<tr><td style="text-align: right;">  4.2</td><td style="text-align: right;">3.5</td><td style="text-align: right;">     2</td><td>[4.0,4.1,4.0,3.9]        </td><td style="text-align: right;">               3.8</td><td style="text-align: right;">      8798</td><td>Collateral (2004)                                 </td><td>Action|Crime|Drama|Thriller               </td></tr>\n<tr><td style="text-align: right;">  4.2</td><td style="text-align: right;">4.0</td><td style="text-align: right;">     2</td><td>[4.0,4.1,3.9]            </td><td style="text-align: right;">               4.0</td><td style="text-align: right;">      6874</td><td>Kill Bill: Vol. 1 (2003)                          </td><td>Action|Crime|Thriller                     </td></tr>\n<tr><td style="text-align: right;">  4.2</td><td style="text-align: right;">4.0</td><td style="text-align: right;">     2</td><td>[4.0,4.1,4.0,4.0,3.9,3.9]</td><td style="text-align: right;">               4.1</td><td style="text-align: right;">     79132</td><td>Inception (2010)                                  </td><td>Action|Crime|Drama|Mystery|Sci-Fi|Thriller</td></tr>\n<tr><td style="text-align: right;">  4.1</td><td style="text-align: right;">3.5</td><td style="text-align: right;">     2</td><td>[4.0,4.0]                </td><td style="text-align: right;">               3.9</td><td style="text-align: right;">     99114</td><td>Django Unchained (2012)                           </td><td>Action|Drama                              </td></tr>\n<tr><td style="text-align: right;">  4.1</td><td style="text-align: right;">4.5</td><td style="text-align: right;">     2</td><td>[4.0,4.1,4.0]            </td><td style="text-align: right;">               4.2</td><td style="text-align: right;">     58559</td><td>Dark Knight, The (2008)                           </td><td>Action|Crime|Drama                        </td></tr>\n<tr><td style="text-align: right;">  4.1</td><td style="text-align: right;">4.5</td><td style="text-align: right;">     2</td><td>[4.0,4.0]                </td><td style="text-align: right;">               4.1</td><td style="text-align: right;">     68157</td><td>Inglourious Basterds (2009)                       </td><td>Action|Drama                              </td></tr>\n<tr><td style="text-align: right;">  4.1</td><td style="text-align: right;">5.0</td><td style="text-align: right;">     2</td><td>[4.0,4.1,4.0]            </td><td style="text-align: right;">               3.9</td><td style="text-align: right;">    106782</td><td>Wolf of Wall Street, The (2013)                   </td><td>Comedy|Crime|Drama                        </td></tr>\n<tr><td style="text-align: right;">  4.1</td><td style="text-align: right;">4.5</td><td style="text-align: right;">     2</td><td>[4.1,4.0,3.9]            </td><td style="text-align: right;">               4.0</td><td style="text-align: right;">     80489</td><td>Town, The (2010)                                  </td><td>Crime|Drama|Thriller                      </td></tr>\n<tr><td style="text-align: right;">  4.0</td><td style="text-align: right;">3.5</td><td style="text-align: right;">     2</td><td>[4.0,4.2,4.1]            </td><td style="text-align: right;"> 

In [ ]:
# GRADED_FUNCTION: sq_dist
# UNQ_C2
def sq_dist(a,b):
    """
    Returns the squared distance between two vectors
    Args:
      a (ndarray (n,)): vector with n features
      b (ndarray (n,)): vector with n features
    Returns:
      d (float) : distance
    """
    ### START CODE HERE ###
    d = np.sum(np.square(a-b))
    ### END CODE HERE ###
    return d

In [ ]:
a1 = np.array([1.0, 2.0, 3.0]); b1 = np.array([1.0, 2.0, 3.0])
a2 = np.array([1.1, 2.1, 3.1]); b2 = np.array([1.0, 2.0, 3.0])
a3 = np.array([0, 1, 0]);       b3 = np.array([1, 0, 0])
print(f"squared distance between a1 and b1: {sq_dist(a1, b1):0.3f}")
print(f"squared distance between a2 and b2: {sq_dist(a2, b2):0.3f}")
print(f"squared distance between a3 and b3: {sq_dist(a3, b3):0.3f}")

squared distance between a1 and b1: 0.000
squared distance between a2 and b2: 0.030
squared distance between a3 and b3: 2.000


In [ ]:
test_sq_dist(sq_dist)

All tests passed!


In [ ]:
input_item_m = tf.keras.layers.Input(shape=(num_item_features))    # input layer
vm_m = item_NN(input_item_m)                                       # use the trained item_NN
vm_m = tf.linalg.l2_normalize(vm_m, axis=1)                        # incorporate normalization as was done in the original model
model_m = tf.keras.Model(input_item_m, vm_m)
model_m.summary()

Model: "model_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_3 (InputLayer)        [(None, 16)]              0         
                                                                 
 sequential_1 (Sequential)   (None, 32)                41376     
                                                                 
 tf.math.l2_normalize_2 (TFO  (None, 32)               0         
 pLambda)                                                        
                                                                 
Total params: 41,376
Trainable params: 41,376
Non-trainable params: 0
_________________________________________________________________


In [ ]:
scaled_item_vecs = scalerItem.transform(item_vecs)
vms = model_m.predict(scaled_item_vecs[:,i_s:])
print(f"size of all predicted movie feature vectors: {vms.shape}")

27/27 [==============================] - 0s 1ms/step
size of all predicted movie feature vectors: (847, 32)


In [ ]:
count = 50  # number of movies to display
dim = len(vms)
dist = np.zeros((dim,dim))

for i in range(dim):
    for j in range(dim):
        dist[i,j] = sq_dist(vms[i, :], vms[j, :])

m_dist = ma.masked_array(dist, mask=np.identity(dist.shape[0]))  # mask the diagonal

disp = [["movie1", "genres", "movie2", "genres"]]
for i in range(count):
    min_idx = np.argmin(m_dist[i])
    movie1_id = int(item_vecs[i,0])
    movie2_id = int(item_vecs[min_idx,0])
    disp.append( [movie_dict[movie1_id]['title'], movie_dict[movie1_id]['genres'],
                  movie_dict[movie2_id]['title'], movie_dict[movie1_id]['genres']]
               )
table = tabulate.tabulate(disp, tablefmt='html', headers="firstrow")

In [ ]:
table

'<table>\n<thead>\n<tr><th>movie1                                  </th><th>genres                                             </th><th>movie2                                        </th><th>genres                                             </th></tr>\n</thead>\n<tbody>\n<tr><td>Save the Last Dance (2001)              </td><td>Drama|Romance                                      </td><td>Mona Lisa Smile (2003)                        </td><td>Drama|Romance                                      </td></tr>\n<tr><td>Wedding Planner, The (2001)             </td><td>Comedy|Romance                                     </td><td>Mr. Deeds (2002)                              </td><td>Comedy|Romance                                     </td></tr>\n<tr><td>Hannibal (2001)                         </td><td>Horror|Thriller                                    </td><td>Final Destination 2 (2003)                    </td><td>Horror|Thriller                                    </td></tr>\n<tr><td>Saving Silverman (Evil Woman) (2001)    </td><td>Comedy|Romance                                     </td><td>Sweetest Thing, The (2002)                    </td><td>Comedy|Romance                                     </td></tr>\n<tr><td>Down to Earth (2001)                    </td><td>Comedy|Fantasy|Romance                             </td><td>Bewitched (2005)                              </td><td>Comedy|Fantasy|Romance                             </td></tr>\n<tr><td>Mexican, The (2001)                     </td><td>Action|Comedy                                      </td><td>Rush Hour 2 (2001)                            </td><td>Action|Comedy                                      </td></tr>\n<tr><td>15 Minutes (2001)                       </td><td>Thriller                                           </td><td>Panic Room (2002)                             </td><td>Thriller                                           </td></tr>\n<tr><td>Enemy at the Gates (2001)               </td><td>Drama                                              </td><td>Whale Rider (2002)                            </td><td>Drama                                              </td></tr>\n<tr><td>Heartbreakers (2001)                    </td><td>Comedy|Crime|Romance                               </td><td>Elektra (2005)                                </td><td>Comedy|Crime|Romance                               </td></tr>\n<tr><td>Spy Kids (2001)                         </td><td>Action|Adventure|Children|Comedy                   </td><td>Scooby-Doo (2002)                             </td><td>Action|Adventure|Children|Comedy                   </td></tr>\n<tr><td>Along Came a Spider (2001)              </td><td>Action|Crime|Mystery|Thriller                      </td><td>Red Dragon (2002)                             </td><td>Action|Crime|Mystery|Thriller                      </td></tr>\n<tr><td>Blow (2001)                             </td><td>Crime|Drama                                        </td><td>25th Hour (2002)                              </td><td>Crime|Drama                                        </td></tr>\n<tr><td>Bridget Jones&#x27;s Diary (2001)            </td><td>Comedy|Drama|Romance                               </td><td>Punch-Drunk Love (2002)                       </td><td>Comedy|Drama|Romance                               </td></tr>\n<tr><td>Joe Dirt (2001)                         </td><td>Adventure|Comedy|Mystery|Romance                   </td><td>Elektra (2005)                                </td><td>Adventure|Comedy|Mystery|Romance                   </td></tr>\n<tr><td>Crocodile Dundee in Los Angeles (2001)  </td><td>Comedy|Drama                                       </td><td>Animal, The (2001)                            </td><td>Comedy|Drama                                       </td></tr>\n<tr><td>Mummy Returns, The (2001)               </td><td>Action|Adventure|Comedy|Thriller                   </td><td>Reign of Fire (2002)                          </td><td>Action|Adventure|Comedy|T